repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java

```diff
@@ -418,6 +433,102 @@ private void traverseEdgesOnDemand(AtlasVertex datasetVertex, boolean isInput, i
         }
     }
 
+    private void traverseEdgesUsingBFS(AtlasVertex baseVertex, AtlasLineageListContext lineageListContext, AtlasLineageListInfo ret) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("traverseEdgesOnDemand");
+
+        Set<String> visitedVertices = new HashSet<>();
+        visitedVertices.add(getGuid(baseVertex));
+        Set<String> skippedVertices = new HashSet<>();
+        Queue<String> traversalQueue = new LinkedList<>();
+
+        enqueueNeighbours(baseVertex, validateEntityTypeAndCheckIfDataSet(getGuid(baseVertex)), lineageListContext, traversalQueue, visitedVertices);
+        int currentDepth = 1;
+
+        while (!traversalQueue.isEmpty() && currentDepth <= lineageListContext.getDepth()) {
+            int entitiesInCurrentDepth = traversalQueue.size();
+
+            for (int i = 0; i < entitiesInCurrentDepth; i++) {
+                String currentGUID = traversalQueue.poll();
+                AtlasVertex currentVertex = AtlasGraphUtilsV2.findByGuid(this.graph, currentGUID);
+                if (Objects.isNull(currentVertex))
+                    throw new AtlasBaseException("Found null vertex during lineage graph traversal for guid: " + currentGUID);
+
+                boolean isDataset = validateEntityTypeAndCheckIfDataSet(currentGUID);
+                if (!lineageListContext.evaluateVertexFilter(currentVertex)) {
+                    enqueueNeighbours(currentVertex, isDataset, lineageListContext, traversalQueue, visitedVertices);
+                    continue;
+                }
+                if (skippedVertices.contains(currentGUID))   // Already skipped vertices due to offset check should not be visited again via any cyclic path
```

Comment:
Shouldn't this skippedVertices check happen before this "if (!lineageListContext.evaluateVertexFilter(currentVertex))" check

Reply:
or maybe we can add this in enqueueNeighbours in this if check -> if (!visitedVertices.contains(getGuid(neighbourVertex)))

Reply:
Please evaluate on your end

Reply:
Moved check to enqueueNeighbours method itself